Controllers/ApplicationController.cs

```diff
@@ -148,19 +95,67 @@ public Task<string> Redeem()
         [HttpGet]
         public Task<string> RedeemPopup()
         {
-            var queryParams = Request.QueryString;
-
             // retrieve CSRF token from browser cookie
-            var csrfToken = Request.Cookies["csrf_token"];
+            var csrfToken = Request.Cookies[CSRF_TOKEN_KEY];
 
+            // check CSRF token and retrieve state and token ID from callback parameters
             return GetMerchantMember()
-                // check CSRF token and retrieve state and token ID from callback parameters
-                .FlatMap(mem => tokenClient.ParseTokenRequestCallbackParams(queryParams, csrfToken.Value)
-                    // get the token and check its validity
-                    .FlatMap(callback => mem.GetToken(callback.TokenId))
-                    // redeem the token at the server to move the funds
-                    .FlatMap(mem.RedeemToken)
-                    .Map(transfer => "Success! Redeemed transfer " + transfer.Id));
+               // check CSRF token and retrieve state and token ID from callback parameters
+               .FlatMap(mem => tokenClient.ParseTokenRequestCallbackUrl(Request.Url.AbsoluteUri, csrfToken.Value)
+                   // get the token and check its validity
+                   .FlatMap(callback => mem.GetToken(callback.TokenId))
+                   // redeem the token at the server to move the funds
+                   .FlatMap(mem.RedeemToken)
+                   .Map(transfer => "Success! Redeemed transfer " + transfer.Id));
+        }
+
+        private static Task<string> InitializeTokenRequestUrl(
```

Comment:
I think this method does way too much. It stores the token request, sets a browser cookie and generates the token request url, so it's really hard to come up with a name that encompasses that. Maybe you can just use a method that converts the queryData into a token request?

Reply:
We followed same format as it is used in java merchant sample. Do we still need to change this

Reply:
Oh interesting... no you're fine then.